src/client/opamSolution.ml

```diff
@@ -138,6 +138,36 @@ let atom_of_name name =
   name, None
 *)
 
+let did_you_mean st atoms = 
+  let open OpamPackage.Set.Op in 
+  let packages =
+    OpamFormula.packages_of_atoms ~disj:false
+      (st.packages ++ st.installed) atoms
+  in
+  let choices name = 
+    let dict = fun yield -> List.iter yield 
+        (OpamPackage.Set.to_list (st.packages ++ st.installed) |> 
+         List.map (fun p -> OpamPackage.name p |> 
```

Comment:
You're transforming the set twice then iterating on it instead of iterating on it directly. That seems wasteful to me.